src/ServiceProviders/RepositoryProviders/GitLab/GitLabProvider.php

```diff
@@ -0,0 +1,246 @@
+        if (!$token) {
+            throw new \Exception('Could not determine authentication token for GitLab serivces. Please set ' . $tokenKey);
+        }
+        $this->setToken($token);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function authenticatedUser()
+    {
+        $userData = $this->gitLabAPI('user');
+        return $userData['username'];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function createRepository($local_site_path, $target, $gitlab_org = '')
+    {
+        $createRepoUrl = "api/v4/projects";
+        $target_org = $gitlab_org;
+        if (empty($gitlab_org)) {
+            $target_org = $this->authenticatedUser();
+            $postData = ['name' => $target];
+        }
+        else {
+            // We need to look up the namespace ID.
+            $group = $this->gitLabAPI('api/v4/groups/' . urlencode($gitlab_org));
+            if (!empty($group)) {
+                $postData = ['name' => $target, 'namespace_id' => $group['id']];
+            }
+            else {
+                $postData = ['name' => $target];
+            }
+        }
+        $target_project = "$target_org/$target";
+
+        // Create a GitLab repository
+        $this->logger->notice('Creating repository {repo}', ['repo' => $target_project]);
+
+        $result = $this->gitLabAPI($createRepoUrl, $postData);
+
+        // Create a git repository. Add an origin just to have the data there
+        // when collecting the build metadata later. We use the 'pantheon'
+        // remote when pushing.
+        // TODO: Do we need to remove $local_site_path/.git? (-n in create-project should obviate this need) We preserve this here because it may be user-provided via --preserve-local-repository
+        if (!is_dir("$local_site_path/.git")) {
+            $this->execGit($local_site_path, 'init');
+        }
+        // TODO: maybe in the future we will not need to set this?
+        $this->execGit($local_site_path, "remote add origin 'git@" . $this->GITLAB_URL . ":{$target_project}.git'");
+
+        return $target_project;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function pushRepository($dir, $target_project)
+    {
+        $this->execGit($dir, 'push --progress https://oauth2:{token}@{gitlab_url}/{target}.git master', ['token' => $this->token(), 'gitlab_url' => $this->GITLAB_URL, 'target' => $target_project], ['token']);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function deleteRepository($project)
+    {
+        $deleteRepoUrl = "api/v4/projects/$project";
+        $this->gitLabAPI($deleteRepoUrl, [], 'DELETE');
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function projectURL($target_project)
+    {
+        return 'https://' . $this->GITLAB_URL . '/' . $target_project;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function commentOnCommit($target_project, $commit_hash, $message)
+    {
+        // @TODO -- Figure out if this is possible in GitLab API.
```

Comment:
Here's a working implementation for this function:

        $url = "api/v4/projects/".urlencode($target_project)
            ."/repository/commits/$commit_hash/comments";
        $data = [ 'note' => $message ];
        $this->gitLabAPI($url, $data);